### PR TITLE
[ISSUE#11659] Develop config query chain of responsibility.

### LINKED ISSUE #11659: using Chain of Responsibility to refactor ConfigQueryRequestHandler and ConfigServletInner
using Chain of Responsibility refactor ConfigQueryRequestHandler  and ConfigServletInner, replacing current serval if-else statements. 

Create ConfigQueryHandler and ConfigQueryHandlerChain,ConfigQueryHandlerChainBuilder.
To exact BetaConfigQueryHandler-beta ,TagConfigQueryHandler-tag,FormalConfigQueryHandler-formal .

ConfigQueryHandlerChainBuilder builder with Handler of ordered name [ beta,tag,formal], if any handler matches this request,handler the request and return , delivery to next handler if not match.

ConfigQueryHandlerChainBuilder may design as a SPI interface ,for extensibility

### PR DESCRIPTION
**Pull request by MatthewAden**
it closes #11659 
## What is the purpose of the change

XXXXX

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java [hunk 1]
```diff
@@ -60,134 +58,136 @@
...
+    public ConfigQueryChainRequest buildChainRequest(ConfigQueryRequest request, RequestMeta meta) {
+        ConfigQueryChainRequest chainRequest = new ConfigQueryChainRequest();
+        
+        String tag = request.getTag();
+        Map<String, String> appLabels = new HashMap<>();
+        appLabels.put(BetaGrayRule.CLIENT_IP_LABEL, meta.getClientIp());
+        if (StringUtils.isNotBlank(tag)) {
+            appLabels.put(TagGrayRule.VIP_SERVER_TAG_LABEL, tag);
+        } else {
+            appLabels = new HashMap<>(meta.getAppLabels());
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java:165]: appLabels.putAll(meta.getAppLabels());

File: config/src/main/java/com/alibaba/nacos/config/server/model/ConfigQueryChainResponse.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    private long lastModified;
+    
+    private ConfigCacheGray matchedGray;
+    
+    private ConfigQueryStatus status;
+    
+    public enum ConfigQueryStatus {
+        BETA,
+        TAG,
+        TAG_NOT_FOUND,
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/model/ConfigQueryChainResponse.java:44]: beta,tag-> gray

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ */
+@Service
+public class ConfigQueryChainService {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigQueryChainService.class);
+    
+    private final ConfigQueryHandlerChain configQueryHandlerChain;
+    
+    public ConfigQueryChainService() {
+        Collection<ConfigQueryHandlerChainBuilder> configQueryHandlerChainBuilders =
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java:46]: ConfigQueryHandlerChainBuilder should  load multi , but use only one  with specific name which's default value is nacos.

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                NacosServiceLoader.load(ConfigQueryHandlerChainBuilder.class);
+        if (CollectionUtils.isEmpty(configQueryHandlerChainBuilders)) {
+            throw new NacosConfigException("No ConfigQueryHandlerChainBuilder found");
+        }
+        ConfigQueryHandlerChainBuilder builder = configQueryHandlerChainBuilders.iterator().next();
+        configQueryHandlerChain = builder.addHandler(new ConfigChainEntryHandler())
+                .addHandler(new GrayRuleMatchHandler())
+                .addHandler(new TagNotFoundHandler())
+                .addHandler(new FormalHandler())
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java:55]: should be extracted to DefaultConfigQueryHandlerChainBuilder with name of 'nacos.'

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/GrayRuleMatchHandler.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * This class represents a gray rule handler in the configuration query processing chain.
+ * It checks if the request matches any gray rules and processes the request accordingly.
+ *
+ * @author Nacos
+ */
+public class GrayRuleMatchHandler extends AbstractConfigQueryHandler {
+    
+    private static final String GRAY_RULE_MATCH_HANDLER = "grayRuleMatchHandler";
+    
+    private ConfigCacheGray matchedGray;
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/GrayRuleMatchHandler.java:42]: Thread not safe

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/TagNotFoundHandler.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.io.IOException;
+
+/**
+ * TagNotFoundHandler Handler.
+ * This class represents tag not found handler in the configuration query processing chain.
+ *
+ * @author Nacos
+ */
+public class TagNotFoundHandler extends AbstractConfigQueryHandler {
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/TagNotFoundHandler.java:31]: change to SpecificTagConfigQueryHandler

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/DefaultConfigQueryHandlerChainBuilder.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultConfigQueryHandlerChainBuilder.class);
+    
+    private ConfigQueryHandler head;
+    
+    private ConfigQueryHandler tail;
+    
+    @Override
+    public ConfigQueryHandlerChain build() {
+        return new ConfigQueryHandlerChain(head);
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/DefaultConfigQueryHandlerChainBuilder.java:39]: ConfigQueryHandlerChain chain =new  ConfigQueryHandlerChain();
        chain.addHandler(new ConfigChainEntryHandler())
                .addHandler(new GrayRuleMatchHandler())
                .addHandler(new TagNotFoundHandler())
                .addHandler(new FormalHandler());
return chain;

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/DefaultConfigQueryHandlerChainBuilder.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        
+        if (head == null) {
+            head = handler;
+            tail = handler;
+        } else {
+            tail.setNextHandler(handler);
+            tail = handler;
+        }
+
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/DefaultConfigQueryHandlerChainBuilder.java:56]: move to ConfigQueryHandlerChain#addHandler

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryHandlerChainBuilder.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+     * @return the configuration query handler chain
+     */
+    ConfigQueryHandlerChain build();
+    
+    /**
+     * Adds a configuration query handler to the chain.
+     * @param handler the handler to be added
+     * @return the current builder instance
+     */
+    ConfigQueryHandlerChainBuilder addHandler(ConfigQueryHandler handler);
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryHandlerChainBuilder.java:38]: move to ConfigQueryHandlerChain#addHandler

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/AbstractConfigQueryHandler.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+ * It implements the {@link ConfigQueryHandler} interface and handles the chaining of handlers.
+ * Subclasses must implement the {@code canHandler} and {@code doHandle} methods to define specific handling logic.
+ *
+ * @author Nacos
+ */
+public abstract class AbstractConfigQueryHandler implements ConfigQueryHandler {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractConfigQueryHandler.class);
+    
+    public ConfigQueryHandler nextHandler;
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/AbstractConfigQueryHandler.java:38]: add prevHandler

**Comment by wuyfee** [comment]
$\color{green}{SUCCESS}$
[DETAILS](https://github.com/alibaba/nacos/actions/runs/12004558892)
 ✅ - docker:   success 
 ✅ - deploy (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-java-test (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-go-test (standalone & cluster):   success 
 ✅ - e2e-cpp-test (standalone & cluster):   success 
 ✅ - e2e-csharp-test (standalone & cluster):   success 
 ✅ - e2e-nodejs-test (standalone & cluster):   success 
 ✅ - e2e-python-test (standalone & cluster):   success 
 ✅ - clean (standalone & cluster & standalone_auth):   success

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryHandlerChain.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import com.alibaba.nacos.config.server.remote.query.handler.ConfigQueryHandler;
+
+import java.io.IOException;
+
+/**
+ * ConfigQueryHandlerChain.
+ * @author Nacos
+ */
+public class ConfigQueryHandlerChain {
+
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryHandlerChain.java:30]: 这些类不应该迁移到service包下，不应该放在remote下

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+     *
+     * @param request the configuration query request object
+     * @return the configuration query response object
+     */
+    public ConfigQueryChainResponse handle(ConfigQueryChainRequest request) {
+        try {
+            return configQueryHandlerChain.handle(request);
+        } catch (Exception e) {
+            LOGGER.error("[Error] Fail to handle ConfigQueryChainRequest", e);
+            return new ConfigQueryChainResponse();
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java:70]: should return  error response with error info

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/ConfigQueryHandler.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+ * This interface defines the standard methods for handling configuration query requests.
+ * @author Nacos
+ */
+public interface ConfigQueryHandler {
+    
+    /**
+     * Gets the name of the handler.
+     * @return The name of the handler.
+     */
+    String getQueryHandlerName();
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/ConfigQueryHandler.java:35]: change to getName()

File: config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java [hunk 14]
```diff
@@ -60,134 +58,136 @@
...
+    }
+    
+    /**
+     * Builds a ConfigQueryChainRequest object.
+     *
+     * @param request the configuration query request
+     * @param meta the request meta
+     * @return the constructed ConfigQueryChainRequest object
+     */
+    public ConfigQueryChainRequest buildChainRequest(ConfigQueryRequest request, RequestMeta meta) {
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java:156]: extract a new spi component which named ConfigQueryChainRequestExtractor, with two method.

extract(HttpServerletRequest request) ,  extract(ConfigQueryRequest request, RequestMeta requestMeta)

File: config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java (hunk 1 above)
Comment (MatthewAden) [reply config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java]: done

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java (hunk 4 above)
Comment (MatthewAden) [reply config/src/main/java/com/alibaba/nacos/config/server/remote/query/ConfigQueryChainService.java]: done

File: config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java [hunk 15]
```diff
@@ -60,134 +58,136 @@
...
+            String pullEvent = resolvePullEventType(chainResponse, request.getTag());
+            LogUtil.PULL_CHECK_LOG.warn("{}|{}|{}|{}", groupKey, clientIp, response.getMd5(), TimeUtils.getCurrentTimeStr());
+            final long delayed = notify ? -1 : System.currentTimeMillis() - response.getLastModified();
+            ConfigTraceService.logPullEvent(dataId, group, tenant, requestIpApp, response.getLastModified(), pullEvent, pullType,
+                    delayed, clientIp, notify, "grpc");
+            
+            return response;
+            
         } catch (Exception e) {
             return ConfigQueryResponse.buildFailResponse(ResponseCode.FAIL.getCode(), e.getMessage());
```
[the remark this answers is not in the document]
Comment (MatthewAden) [reply config/src/main/java/com/alibaba/nacos/config/server/remote/ConfigQueryRequestHandler.java:137]: done

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/AbstractConfigQueryHandler.java (hunk 10 above)
Comment (MatthewAden) [reply config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/AbstractConfigQueryHandler.java]: ignore

File: config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/ConfigQueryHandler.java (hunk 13 above)
Comment (MatthewAden) [reply config/src/main/java/com/alibaba/nacos/config/server/remote/query/handler/ConfigQueryHandler.java]: done

File: config/src/main/java/com/alibaba/nacos/config/server/service/query/ConfigQueryChainRequestExtractor.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param request  the HTTP request object
+     * @param dataId   the configuration data ID
+     * @param group    the configuration group
+     * @param tenant   the tenant information
+     * @param tag      the tag
+     * @param clientIp the client IP address
+     * @return the extracted configuration query chain request
+     */
+    ConfigQueryChainRequest extract(HttpServletRequest request, String dataId, String group, String tenant, String tag, String clientIp);
```
Comment (shiyiyue1102) [inline config/src/main/java/com/alibaba/nacos/config/server/service/query/ConfigQueryChainRequestExtractor.java:50]: simply extractor method， keep  HttpServletRequest request only

**Comment by wuyfee** [comment]
$\color{green}{SUCCESS}$
[DETAILS](https://github.com/alibaba/nacos/actions/runs/12112454078)
 ✅ - docker:   success 
 ✅ - deploy (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-java-test (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-go-test (standalone & cluster):   success 
 ✅ - e2e-cpp-test (standalone & cluster):   success 
 ✅ - e2e-csharp-test (standalone & cluster):   success 
 ✅ - e2e-nodejs-test (standalone & cluster):   success 
 ✅ - e2e-python-test (standalone & cluster):   success 
 ✅ - clean (standalone & cluster & standalone_auth):   success

**Comment by wuyfee** [comment]
$\color{green}{SUCCESS}$
[DETAILS](https://github.com/alibaba/nacos/actions/runs/12133834611)
 ✅ - docker:   success 
 ✅ - deploy (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-java-test (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-go-test (standalone & cluster):   success 
 ✅ - e2e-cpp-test (standalone & cluster):   success 
 ✅ - e2e-csharp-test (standalone & cluster):   success 
 ✅ - e2e-nodejs-test (standalone & cluster):   success 
 ✅ - e2e-python-test (standalone & cluster):   success 
 ✅ - clean (standalone & cluster & standalone_auth):   success

**Comment by wuyfee** [comment]
$\color{green}{SUCCESS}$
[DETAILS](https://github.com/alibaba/nacos/actions/runs/12142216837)
 ✅ - docker:   success 
 ✅ - deploy (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-java-test (standalone & cluster & standalone_auth):   success 
 ✅ - e2e-go-test (standalone & cluster):   success 
 ✅ - e2e-cpp-test (standalone & cluster):   success 
 ✅ - e2e-csharp-test (standalone & cluster):   success 
 ✅ - e2e-nodejs-test (standalone & cluster):   success 
 ✅ - e2e-python-test (standalone & cluster):   success 
 ✅ - clean (standalone & cluster & standalone_auth):   success

**Comment by wuyfee** [comment]
$\color{red}{FAILURE}$
[DETAILS](https://github.com/alibaba/nacos/actions/runs/12151531369)
 ✅ - docker:   success 
 ✅ - deploy (standalone & cluster & standalone_auth):   success 
 ❌ - e2e-java-test (standalone & cluster & standalone_auth):   failure 
 ✅ - e2e-go-test (standalone & cluster):   success 
 ✅ - e2e-cpp-test (standalone & cluster):   success 
 ✅ - e2e-csharp-test (standalone & cluster):   success 
 ✅ - e2e-nodejs-test (standalone & cluster):   success 
 ✅ - e2e-python-test (standalone & cluster):   success 
 ✅ - clean (standalone & cluster & standalone_auth):   success

**Review by shiyiyue1102 — APPROVED** [review]
approve

**Pull request merged**